grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/Atom.java

```diff
@@ -151,24 +151,22 @@ public int baseResolutionPriority(){
     /**
      * @return set of potentially applicable rules - does shallow (fast) check for applicability
      */
-    private Set<Rule> getPotentialRules(){
-        OntologyConcept ontologyConcept = getOntologyConcept();
-        return ontologyConcept != null ?
-                ontologyConcept.subs().flatMap(OntologyConcept::getRulesOfConclusion).collect(Collectors.toSet()) :
-                ReasonerUtils.getRules(graph());
+    private Stream<Rule> getPotentialRules(){
+        return new RuleGraph(graph()).getRulesWithType(getOntologyConcept());
     }
 
     /**
      * @return set of applicable rules - does detailed (slow) check for applicability
      */
-    public Set<InferenceRule> getApplicableRules() {
+    public Stream<InferenceRule> getApplicableRules() {
```

Comment:
Collecting into a set just to convert into a stream is slightly inefficient. How about: 

```
if (applicableRules == null) {
    return getPotentialRules()
                   .map(rule -> new InferenceRule(rule, graph()))
                   .filter(this::isRuleApplicable)
                   .map(r -> r.rewriteToUserDefined(this))
} else {
    return applicableRules.stream();
}
```

In the above case you avoid a potential serialisation into a collection just to be converted into a stream again